flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -174,6 +181,15 @@ public void registerAll(
         }
     }
 
+    @Override
+    public void registerAllAfterRestored(CompletedCheckpoint checkpoint, RestoreMode mode) {
+        registerAll(checkpoint.getOperatorStates().values(), checkpoint.getCheckpointID());
+        if (mode != RestoreMode.CLAIM) {
```

Comment:
Why only `CLAIM` mode does not need to update the `highestRetainCheckpointID`? I think this deserve a description.

Reply:
In `CLAIM` mode, the state of the checkpoints that the job is recovering from can be deleted (must be deleted).
I'll add this comment.